### ISSUE TITLE
Resize to fit fluid container

### ISSUE THREAD

**Issue by tompazourek** [imgix/imgix.js#117]
I have a fluid-width container that has many different sizes depending on how it collapses with regards to content around it. The size of the container does not always depend just on the viewport breakpoints, but also it depends on the content around it.

Is there some way how I could use imgix.js to react to how the container's size changes and fit the image inside with some good dimensions?

The issue kind of resembles the media queries vs. container queries thing a bit.

Do you have some ideas or recommendations how I could deal with those more fluid layouts while still loading only the image sizes actually needed?

Thanks

**Comment by jayeb** [comment]
Hey Tom,

imgix.js is actually kind of agnostic to your problem here--it's not going to help or hinder your situation at all. This library is concerned exclusively with populating your `<img>` tag's `srcset` attribute, regardless of how the image is sized in your page's layout. We populate the `srcset` with a huge array of reasonable image URLs, and let the browser determine which of them to use. The good news here is that this array of URLs is the same no matter what size your image is being rendered at. Put another way: no matter how big your browser is, the URL for a 300px-wide version of your image is still `example.imgix.net/image.jpg?w=300`.

The attribute of concern in the situation you've described is the `sizes` attribute, which imgix.js doesn't touch. If you make sure the value of `sizes` reflects the intended width of your `<img>` element, the browser will then select the closest image URL from the `srcset` attribute (helpfully populated by imgix.js) and download & display that value to your end-user. So the problem you're really looking to solve here is "how can I make sure my image's `sizes` attribute has an accurate size value?"

The easiest way to do this is to just define the `sizes` value statically. Depending on your layout this might not be possible (and it sounds like maybe it's not), but keep in mind that you can use `vw` values in here if your image is sized based on the viewport's width.

If you can't define your image's size statically, then you'll have to leverage some relatively-simple Javascript to get the job done. First and foremost, you'll want to turn off imgix.js's auto-initialization feature. By default, imgix.js tries to run as soon as possible on page load. If the `<img>` tags it's trying to run on don't have a `sizes` attribute defined at load time (which yours won't), it automatically sets the value to `100vw`. Once imgix.js is done setting the `srcset` on the `<img>`, the browser will begin downloading a file under the assumption that the `<img>` is meant to fill the entire viewport. This will end up being a huge waste of time and bandwidth for your end-user, so we'll want to turn this off with a `<meta>` tag in your page's head. There are [instructions to do this](https://github.com/imgix/imgix.js#disabling-auto-initialization) on the README.

To manually set your `sizes` attribute, you'll want to wait until the page is done painting for the first time, then measure your `<img>` tag and update its `sizes` attribute to reflect its current width in the layout. Once you do this, call `imgix.init()` and imgix.js will run, filling in the `<img>` tag's `srcset` attribure. As soon as the browser sees a `sizes` attribute and a `srcset` attribute on the same image, it has all the info it needs to select an image URL and kick off the download.

If you need to be able to react to viewport size changes after the initial page load, just attach a handler to the window's `resize` event and re-calculate and re-apply the `sizes` attribute on your image every time it's fired.

Apologies for my long-windedness, but I hope this helps! I'm going to close this ticket, but if you have any follow-up questions feel free to keep the conversation going.

**Comment by tompazourek** [comment]
Hi Jason,

Thanks a lot for the comprehensive response, I appreciate your help a lot.

If I understand it correctly, the best way of tackling this would be to always compute the desired size of the image by JS and then set this image size to the `sizes` attribute of imgix.

I'm just curious here: If I set the `sizes` attribute of imgix to 600px and the image is loaded and put in cache, and then later after some resizing has been done, I recompute the image size and set the `sizes` attribute by JS to 500px. Will imgix still be able to use the old 600px image that was already loaded and is in cache? Or will it try to load the 500px image now, because I changed the `sizes` attribute?

When I change the `sizes` attribute, do I need to re-initialize imgix? Is it possible to initialize just a specific single image with the `.init()` method?

**Comment by jayeb** [comment]
> If I understand it correctly, the best way of tackling this would be to always compute the desired size of the image by JS and then set this image size to the sizes attribute of imgix.

Yup! You've got it.

> Will imgix still be able to use the old 600px image that was already loaded and is in cache? Or will it try to load the 500px image now, because I changed the sizes attribute?

This decision isn't up to imgix.js at all--remember, once the `srcset` attribute has been filled in, imgix.js's work is done. From this point forward, it's up to the browser to choose which image to use. I believe at this point most browsers should be smart enough to not reload the image when it gets scaled down, but the `srcset`/`sizes` spec was purposefully written to be ambiguous here, choosing to leave the choice of which image to load at any given time up to the browser vendor.

> When I change the sizes attribute, do I need to re-initialize imgix?

Nope! Once imgix.js has filled in the `srcset` attribute for any given image element, its work is done. The extensive `srcset` list it generates contains more than enough image URLs for the browser to choose from no matter what value you provide in `sizes`.

> Is it possible to initialize just a specific single image with the .init() method?

It is, but you shouldn't need to worry about this. A call to `imgix.init()` is **idempotent**, meaning that its work is only done once no matter how many times it's run (although you can [override this](https://github.com/imgix/imgix.js#imgix-init-idempotency) if needed). If you're adding `<img>` tags to the page after your initial `imgix.init()` call, you can just call `imgix.init()` again and be confident that everything will work: the new image elements will get their `srcset` attributes filled in correctly, and old images will remain as they were without any wasted effort.

**Comment by tompazourek** [comment]
@jayeb Thanks again for your insights, lot of useful information there. I'll have to think it through properly to get it completely before I attempt to actually implement it. The responsive images are still a bit new to me.

I just have a quick question about the way how imgix.js generates the `srcset`. Can you summarize the method imgix.js uses to transform the `ix-src` to the list of `srcset`? From what I've read and seen in the documentation, it seems that it does all sizes from 100w to 2560w in 100w increments while keeping the aspect ratio. How is retina handled here? Is it always in increments of 100w with the 100w minimum and 2560w maximum? I'm guessing the 100w increments are to better leverage browser caching, or what's the reason there? Sorry if the question is a bit stupid.

**Comment by jayeb** [comment]
> Can you summarize the method imgix.js uses to transform the ix-src to the list of srcset? From what I've read and seen in the documentation, it seems that it does all sizes from 100w to 2560w in 100w increments while keeping the aspect ratio

Yup, you've got it. You can see the full list of sizes we use [here](https://github.com/imgix/imgix.js/blob/master/src/targetWidths.js), but in a nutshell the rule is "the full-screen width of a bunch of popular devices (multiplied by that device's DPR), plus a width for every 100-pixel increment between 100 and 2560." The aspect ratio doesn't actually matter here: the browser expects all of the URLs listed in a `srcset` to list either a width *or* a height (we use width, which is far more popular), but not both. Aspect ratio isn't particularly important.

> How is retina handled here?

The way imgix.js is built, we don't even worry about device-pixel-ratio at all. Again, all imgix.js does is build a `srcset` with a comprehensive list of image URLs at a wide variety of widths. It's up to the browser to decide which image to download, based on what it knows about the device it's running on and the size of the image element it's trying to fill (which it knows because of the `sizes` attribute. So if the browser knows the `<img>` it's trying to fill is 200px wide and the DPR of the device is 2, it's going to look through the `srcset` list for the image URL closest to 400px wide and download that one. Likewise, a browser running on a DPR 3 device will look through the same list and select the image URL closest to 600px wide and download that one.

> I'm guessing the 100w increments are to better leverage browser caching, or what's the reason there?

You're close--it's not the browser cache that matters here as much as it is the CDN cache. By having a wide-but-not-infinite number of possible images to download, we're trying to find the sweet-spot between giving the browser an image that's pretty close to the perfect size (thereby reducing the number of wasted bytes it has to download and resize locally) and giving the browser an image that's already rendered and cached in *our* system (thereby speeding up the request time by a huge factor).

**Comment by tompazourek** [comment]
@jayeb: That's great. Thanks a lot for your time, Jason, you've made a lot of things much clearer for me. 😄 I think most of my misunderstanding came from not knowing which part is handled by imgix.js and which is built-in inside the browser. I went through the source code and found out it isn't so complicated and magical as I thought originally.

I think it might be because last time I was looking into imgix.js, it was still version 2, and I didn't realize that so much has changed since then. I've read through #98 and I see the points made there. It all makes sense now. Thank you!
